src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.cs

```diff
@@ -0,0 +1,368 @@
+            return Assembly.Load(assemblyName, _nativeAssemblyLoadContext);
+        }
+
+        // These methods load assemblies into the current AssemblyLoadContext 
+        // They may be used in the implementation of an AssemblyLoadContext derivation
+        public Assembly LoadFromAssemblyPath(string assemblyPath)
+        {
+            if (assemblyPath == null)
+            {
+                throw new ArgumentNullException(nameof(assemblyPath));
+            }
+
+            if (PathInternal.IsPartiallyQualified(assemblyPath))
+            {
+                throw new ArgumentException(SR.Argument_AbsolutePathRequired, nameof(assemblyPath));
+            }
+
+            lock (_unloadLock)
+            {
+                VerifyIsAlive();
+
+                return InternalLoadFromPath(assemblyPath, null);
+            }
+        }
+
+        public Assembly LoadFromNativeImagePath(string nativeImagePath, string assemblyPath)
+        {
+            if (nativeImagePath == null)
+            {
+                throw new ArgumentNullException(nameof(nativeImagePath));
+            }
+
+            if (PathInternal.IsPartiallyQualified(nativeImagePath))
+            {
+                throw new ArgumentException(SR.Argument_AbsolutePathRequired, nameof(nativeImagePath));
+            }
+
+            if (assemblyPath != null && PathInternal.IsPartiallyQualified(assemblyPath))
+            {
+                throw new ArgumentException(SR.Argument_AbsolutePathRequired, nameof(assemblyPath));
+            }
+
+            lock (_unloadLock)
+            {
+                VerifyIsAlive();
+
+                return InternalLoadFromPath(assemblyPath, nativeImagePath);
+            }
+        }        
+
+        public Assembly LoadFromStream(Stream assembly)
+        {
+            return LoadFromStream(assembly, null);
+        }
+
+        public Assembly LoadFromStream(Stream assembly, Stream assemblySymbols)
+        {
+            if (assembly == null)
+            {
+                throw new ArgumentNullException(nameof(assembly));
+            }
+
+            if (assembly.Length <= 0)
+            {
+                throw new BadImageFormatException(SR.BadImageFormat_BadILFormat);
+            }
+
+            lock (_unloadLock)
+            {
+                VerifyIsAlive();
+
+                int iAssemblyStreamLength = (int) assembly.Length;
+
+                // Allocate the byte[] to hold the assembly
+                byte[] arrAssembly = new byte[iAssemblyStreamLength];
+
+                // Copy the assembly to the byte array
+                assembly.Read(arrAssembly, 0, iAssemblyStreamLength);
+
+                // Get the symbol stream in byte[] if provided
+                byte[] arrSymbols = null;
+                if (assemblySymbols != null)
+                {
+                    var iSymbolLength = (int) assemblySymbols.Length;
+                    arrSymbols = new byte[iSymbolLength];
```

Comment:
This is pretty inefficient, left over from times when this was wrapping `Load(byte[])`. This whole path makes multiple unnecessary copies of the assembly. I would love to fix that.

If you are implementing this for Mono from scratch anyway, I would make this read directly to the final piece of memory that you need the bits to be, in non-shared partition.

Reply:
I agree that the stream buffering is not needed but we also have byte[] based unmanaged implementation which this code will call. I can remove this section but maybe keeping it shared would help with sharing the better stream version in the future.

I have subsequent PR which removes one level of copying ready to PR after this change.